### PR TITLE
Simplify and improve code generation

### PR DESCRIPTION
**Pull request by FabioLuporini**
Padding is now entirely supported by COFFEE, so there is no need to pad the buffer arrays in the wrapper code generation anymore.

As a consequence, code generation in host.py is greatly simplified/improved.

Note that code generation from ASTs to C is now postponed until the invocation of the ``compile`` method in ``JITModules``.

This depends on PR #36 in COFFEE.

### CODE REVIEW

**Comment by kynan** [comment]
From a cursory glance this looks sane.

**Comment by kynan** [comment]
Any thoughts @wence-?

File: pyop2/host.py [hunk 1]
```diff
@@ -651,6 +645,14 @@ def compile(self, argtypes=None, restype=None):
         strip = lambda code: '\n'.join([l for l in code.splitlines()
                                         if l.strip() and l.strip() != ';'])
 
+        # Attach semantical information to the kernel's AST
+        if self._kernel._ast:
+            fundecl = ast_get_fun_decls(self._kernel._ast)
+            if fundecl:
+                for arg, f_arg in zip(self._args, fundecl.args):
+                    if arg._uses_itspace and arg._is_INC:
+                        f_arg.pragma = ast.WRITE
```
Comment (wence-) [inline pyop2/host.py]: A comment here explaining what's going on might be good!.

**Comment by wence-** [comment]
This looks broadly sane.  Does it depend on coneoproject/COFFEE#36 landing simultaneously?

File: pyop2/host.py (hunk 1 above)
Comment (FabioLuporini) [inline pyop2/host.py]: I've changed this piece a bit in another branch I have, so I'll comment it properly there once "ready"

**Comment by FabioLuporini** [comment]
Yes, they depend on each other

**Comment by FabioLuporini** [comment]
Can we merge this and the corresponding coffee changes?

**Comment by wence-** [comment]
This doesn't merge cleanly, so no.

**Comment by wence-** [comment]
I have updated this branch on top of current master.  Please check it.

**Comment by FabioLuporini** [comment]
So there were a few conflicts before your rebase (mostly due to code gen in host.py). Don't know what were they due to. I'll see if tests pass

**Comment by FabioLuporini** [comment]
So tests pass, if you feel the rebase you've done is OK, for me we can merge

**Pull request merged**